src/backend/joanie/core/flows/order.py

```diff
@@ -25,6 +25,135 @@ def _set_order_state(self, value):
     def _get_order_state(self):
         return self.instance.state
 
+    def _can_be_state_completed_from_assigned(self):
+        """
+        An order state can be set to completed if all installments
+        are completed.
+        """
+        return self.instance.is_free and (
+            not self.instance.has_contract or self.instance.contract.student_signed_on
+        )
+
+    def _can_be_state_to_sign_and_to_save_payment_method(self):
+        """
+        An order state can be set to to_sign_and_to_save_payment_method if the order is free
+        and has a contract but no payment method.
+        """
+        return (
+            not self.instance.is_free
+            and not self.instance.has_payment_method
+            and self.instance.has_contract
+        )
+
+    def _can_be_state_to_save_payment_method(self):
+        """
+        An order state can be set to to_save_payment_method if the order is not free
+        and has no payment method and no contract.
+        """
+        return (
+            not self.instance.is_free
+            and not self.instance.has_payment_method
+            and not self.instance.has_contract
+        )
+
+    def _can_be_state_to_sign(self):
+        """
+        An order state can be set to to_sign if the order is free
+        or has a payment method and a contract.
+        """
+        return (
+            self.instance.is_free or self.instance.has_payment_method
+        ) and self.instance.has_contract
```

Comment:
As discussed, an order should be to sign if there is a contract and this one has not been yet signed by the learner.

`order.has_unsigned_contract` 😉